pipeline.go

```diff
@@ -26,66 +26,70 @@ func init() {
 
 // Pipeline stores all definitions and settings regarding a deployment.
 type Pipeline struct {
-	Definition  PipelineDefinition
-	Environment PipelineEnvironment
+	Definition  *PipelineDefinition
+	Environment *PipelineEnvironment
 	NetworkName string
 }
 
-// PipelineDefinition stores docker-compose services and gantry steps.
-type PipelineDefinition struct {
-	Steps        StepList    `json:"steps"`
-	Services     ServiceList `json:"services"`
-	pipelines    *Pipelines
-	IgnoredSteps types.StringSet
+// NewPipeline creates a new Pipeline from given files which ignores the
+// existence of steps with names provided in ignoreSteps.
+func NewPipeline(definitionPath, environmentPath string, ignoredSteps types.StringSet) (*Pipeline, error) {
+	p := &Pipeline{}
+	def, err := NewPipelineDefinition(definitionPath)
+	if err != nil {
+		return nil, err
+	}
+	def.IgnoredSteps = ignoredSteps
+	p.Definition = def
+	p.Environment = NewPipelineEnvironment()
+	if p.checkRequireEnvironment() {
+		err := p.Environment.Load(environmentPath)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if err = p.Environment.ApplyTo(def); err != nil {
+		return nil, err
+	}
+	return p, nil
 }
 
-// Pipelines calculates and verifies dependencies and ordering for steps
-// defined in the PipelineDefinition p.
-func (p *PipelineDefinition) Pipelines() (*Pipelines, error) {
-	if p.pipelines == nil {
-		steps := make(map[string]Step, 0)
-		// Verfiy steps
-		for name, step := range p.Steps {
-			if _, ignore := p.IgnoredSteps[name]; ignore {
-				continue
-			}
-			if val, ok := steps[name]; ok {
-				return nil, fmt.Errorf("Redeclaration of step '%s'", val.Name)
-			}
-			for ignored := range p.IgnoredSteps {
-				delete(step.After, ignored)
-				delete(step.DependsOn, ignored)
-			}
-			steps[name] = step
-		}
-		// Verify services
-		for name, step := range p.Services {
-			if _, ignore := p.IgnoredSteps[name]; ignore {
-				continue
-			}
-			if val, ok := steps[name]; ok {
-				return nil, fmt.Errorf("Redeclaration of step '%s'", val.Name)
-			}
-			for ignored := range p.IgnoredSteps {
-				delete(step.After, ignored)
-				delete(step.DependsOn, ignored)
+func (p *Pipeline) checkRequireEnvironment() bool {
+	return false
```

Comment:
Until there is actual functionality I wouldn't introduce a method for it.

Reply:
In a temporary version some logic existed here, will be removed